sql: implement the client_min_messages session var

otan opened the pull request:
Resolves: https://github.com/cockroachdb/cockroach/issues/44711

PostgreSQL supports notices of varying severity to display to the user.
This is achieved through the `client_min_messages` session var. The
strings it accepts is different than the Severity value.

As such, we create a new pgnotice package that has a `DisplaySeverity`
setting that encodes this value. Severities from pgerror will get
translated into a DisplaySeverity and compared to the
`client_min_messages` value before displaying.

In addition, we have created the `pgnotice` package which
contains the Severity constants as well as moving `pgerror.Noticef` to
`pgnotice.Newf`.

Release note (sql change): Users now have the ability to control which
level of NOTICE output with the session variable `client_min_messages`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46650)
<!-- Reviewable:end -->

knz reviewed (commented):
So I appreciate the increase in functionality performed here. I'd argue we don't have much evidence that this was in demand by users/customers, but if it scratched your itch, so be it it.

In the meantime, I notice that you took a shortcut. The severity field should definitely be part of the `pgerror.Error` object. After all, it is part of the error object in `lib/pq` and `pgx` too. We never really properly thought about it, but the moment you introduce the notion of severity in CockroachDB this should _definitely_ be the place to add it.

However, changing the protobuf at this point of the release cycle would be a bit too much movement for the stability period. Therefore, I encourage you to proceed as follows:

1) issue a PR immediately (separate from this one) that removes the single-argument form of `crdb_internal.notice`, ad gives it a first "severity" argument that would be _currently ignored_. We'd then release 20.1 with that interface.

2) iterate on this current PR further and put the severity in the right place. We'd then have it merge after the release branch is cut.

> Reviewed 24 of 24 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46650#-:-M3fEeCI4-FRnfZjo_pz:b-s0fd1j)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/pgwire/pgnotice/severity.go, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46650#-M3fFJEB2OXCmA0ZVAle:-M3fFJEB2OXCmA0ZVAlf:b-vq122) ([raw file](https://github.com/cockroachdb/cockroach/blob/240ad2e14559c75a91fc2bf869dfc1c3e453b472/pkg/sql/pgwire/pgnotice/severity.go#L25)):*
> ```Go
> 
> const (
> 	// SeverityError is a Severity value allowing all notices
> ```

Please add a comment above this that explains that the ordering is particular and needs to be preserved for compatibility with PostgreSQL.
___
*[pkg/sql/pgwire/pgnotice/severity.go, line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46650#-M3fFQ3GCUmNDa28MK2S:-M3fFQ3GCUmNDa28MK2T:bvavabh) ([raw file](https://github.com/cockroachdb/cockroach/blob/240ad2e14559c75a91fc2bf869dfc1c3e453b472/pkg/sql/pgwire/pgnotice/severity.go#L51)):*
> ```Go
> 	// SeverityDebug5 is a Severity value allowing all notices
> 	// of value <= SeverityDebug5 to display.
> 	SeverityDebug5 Severity = iota
> ```

I don't understand what a `= iota` does here. Doesn't it restart the numbering at 0? Isn't that a problem to have Debug5 = 0 while Debug4 = 7?
___
*[pkg/sql/pgwire/pgnotice/severity.go, line 100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46650#-M3fFcKl7ffjs8JSvqJj:-M3fFcKl7ffjs8JSvqJk:bmkx2tf) ([raw file](https://github.com/cockroachdb/cockroach/blob/240ad2e14559c75a91fc2bf869dfc1c3e453b472/pkg/sql/pgwire/pgnotice/severity.go#L100)):*
> ```Go
> 
> // WireString returns the value to return over pgwire.
> // TODO(otan): we should have the error proto have a "Severity" field,
> ```

Why don't you make this happen then? (See my comment at top)


<!-- Sent from Reviewable.io -->

otan commented:
seems like we're cutting today anyway so i'll put this on the backburner for a bit.

otan commented:
(closing for now!)

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46650)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/sql/pgwire/pgnotice/severity.go, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46650#-M3fFJEB2OXCmA0ZVAle:-M5TwN1a583SA1SHzLJ3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/240ad2e14559c75a91fc2bf869dfc1c3e453b472/pkg/sql/pgwire/pgnotice/severity.go#L25)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please add a comment above this that explains that the ordering is particular and needs to be preserved for compatibility with PostgreSQL.
</blockquote></details>

Done.
___
*[pkg/sql/pgwire/pgnotice/severity.go, line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46650#-M3fFQ3GCUmNDa28MK2S:-M5TwNk89EhwLtieVljF:b-relee4) ([raw file](https://github.com/cockroachdb/cockroach/blob/240ad2e14559c75a91fc2bf869dfc1c3e453b472/pkg/sql/pgwire/pgnotice/severity.go#L51)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't understand what a `= iota` does here. Doesn't it restart the numbering at 0? Isn't that a problem to have Debug5 = 0 while Debug4 = 7?
</blockquote></details>

Done. (bug)
___
*[pkg/sql/pgwire/pgnotice/severity.go, line 100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46650#-M3fFcKl7ffjs8JSvqJj:-M5TwPgr3q4pRzLpqSC8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/240ad2e14559c75a91fc2bf869dfc1c3e453b472/pkg/sql/pgwire/pgnotice/severity.go#L100)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Why don't you make this happen then? (See my comment at top)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with nit. Also you need to rebase to pick up the TC changes.

> Reviewed 25 of 25 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46650#-:-M5WcVVb9dnuldqCt9eg:b-95m2d3)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/sem/builtins/builtins.go, line 3393 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46650#-M5WcKGv56Ex76fnfmEM:-M5WcKGv56Ex76fnfmEN:b-b4eli1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad3fb055fa9a58f2751011f8fe2fedab85878d51/pkg/sql/sem/builtins/builtins.go#L3393)):*
> ```Go
> 				if ctx.ClientNoticeSender == nil {
> 					return nil, errors.AssertionFailedf("notice sender not set")
> 				}
> ```

Can you extract this code into a separate function and call that function from both overloads. Thanks


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46650)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz)

___
*[pkg/sql/sem/builtins/builtins.go, line 3393 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46650#-M5WcKGv56Ex76fnfmEM:-M5YRQJ59LjjZhq4r_ev:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad3fb055fa9a58f2751011f8fe2fedab85878d51/pkg/sql/sem/builtins/builtins.go#L3393)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you extract this code into a separate function and call that function from both overloads. Thanks
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
sql: implement the client_min_messages session var

PostgreSQL supports notices of varying severity to display to the user.
This is achieved through the `client_min_messages` session var. The
strings it accepts is different than the Severity value.

As such, we create a new pgnotice package that has a `DisplaySeverity`
setting that encodes this value. Severities from pgerror will get
translated into a DisplaySeverity and compared to the
`client_min_messages` value before displaying.

In addition, we have created the `pgnotice` package which
contains the Severity constants as well as moving `pgerror.Noticef` to
`pgnotice.Newf`.

Release note (sql change): Users now have the ability to control which
level of NOTICE output with the session variable `client_min_messages`.

otan commented:
tftr!

bors r=knz

otan commented:
flake

bors r=knz
